src/lib/inclusion_user_callbacks.ts

```diff
@@ -0,0 +1,88 @@
+import { InclusionGrant, InclusionUserCallbacks } from "zwave-js";
+import { createDeferredPromise } from "alcalzone-shared/deferred-promise";
+import { Client, ClientsController } from "./server";
+
+export const inclusionUserCallbacks = (
+  clientsController: ClientsController,
+  client?: Client
+): InclusionUserCallbacks => {
+  return {
+    grantSecurityClasses: (
+      requested: InclusionGrant
+    ): Promise<InclusionGrant | false> => {
+      clientsController.grantSecurityClassesPromise = createDeferredPromise();
+      clientsController.grantSecurityClassesPromise.catch(() => {});
+      clientsController.grantSecurityClassesPromise.finally(() => {
+        if (clientsController.grantSecurityClassesPromise !== undefined) {
+          delete clientsController.grantSecurityClassesPromise;
+        }
+      });
+      if (client !== undefined) {
+        client.sendEvent({
+          source: "controller",
+          event: "grant security classes",
+          requested: requested as any,
+        });
+      } else {
+        clientsController.clients.forEach((client) => {
+          if (client.isConnected && client.receiveEvents) {
+            client.sendEvent({
+              source: "controller",
+              event: "grant security classes",
+              requested: requested as any,
+            });
+          }
+        });
+      }
+
+      return clientsController.grantSecurityClassesPromise;
+    },
+    validateDSKAndEnterPIN: (dsk: string): Promise<string | false> => {
+      clientsController.validateDSKAndEnterPinPromise = createDeferredPromise();
+      clientsController.validateDSKAndEnterPinPromise.catch(() => {});
+      clientsController.validateDSKAndEnterPinPromise.finally(() => {
+        if (clientsController.validateDSKAndEnterPinPromise != undefined) {
+          delete clientsController.validateDSKAndEnterPinPromise;
+        }
+      });
+      if (client !== undefined) {
+        client.sendEvent({
+          source: "controller",
+          event: "validate dsk and enter pin",
+          dsk,
+        });
+      } else {
+        clientsController.clients.forEach((client) => {
+          if (client.isConnected && client.receiveEvents) {
+            client.sendEvent({
+              source: "controller",
+              event: "validate dsk and enter pin",
+              dsk,
+            });
+          }
+        });
+      }
+      return clientsController.validateDSKAndEnterPinPromise;
+    },
+    abort: (): void => {
+      // settle the promises to ensure finally is triggered for the cleanup.
+      clientsController.grantSecurityClassesPromise?.resolve(false);
+      clientsController.validateDSKAndEnterPinPromise?.resolve(false);
```

Comment:
I would remove those two lines. `abort` is called by the driver at a point it has timed out waiting for the other promises to resolve.

Reply:
so what happens to the promises?

Reply:
I guess we discard them and then the commands will throw an error if a client tries to complete the flow after the timeout

Reply:
OK so I need to replace resolving them with deleting them. Will do that when i get home